# Segmenting and Clustering Neighborhoods in Toronto

### 1. Import Library

In [227]:
import numpy as np
import pandas as pd
import requests

from bs4 import BeautifulSoup

import json
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported.')

Libraries imported.


### 2. Scrape Data

In [228]:
# scrape the Wikipedia page to obtain the raw data
url= 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
data=requests.get(url).text
#data

In [229]:
# use BeautifulSoup to get process on web text
soup = BeautifulSoup(data,'html5lib')
#print(soup.prettify())

In [230]:
table = soup.find('table')
table_rows = table.find_all('tr')

l = []
for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.strip() for tr in td if tr.text.strip()]
    if row:
        l.append(row)

In [231]:
# Create dataframe
df = pd.DataFrame(l,columns=["PostalCode", "Borough", "Neighbourhood"])
df.head(10)

PostalCode           Borough                                 Neighbourhood
0        M1A      Not assigned                                          None
1        M2A      Not assigned                                          None
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
7        M8A      Not assigned                                          None
8        M9A         Etobicoke                              Islington Avenue
9        M1B       Scarborough                               Malvern / Rouge

### 3. Clean Data

In [232]:
df.describe()

PostalCode       Borough Neighbourhood
count         180           180           103
unique        180            11            98
top           M4P  Not assigned     Downsview
freq            1            77             4

Only process cells that have an assigned Borough. Ignore cells with a borough that is Not assigned.

In [233]:
df=df[df.Borough != 'Not assigned']
df.head(10)

PostalCode           Borough                                 Neighbourhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
8         M9A         Etobicoke                              Islington Avenue
9         M1B       Scarborough                               Malvern / Rouge
11        M3B        North York                                     Don Mills
12        M4B         East York              Parkview Hill / Woodbine Gardens
13        M5B  Downtown Toronto                     Garden District / Ryerson

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

#### Evaluating for Missing Data
As the output shows: after excluding the rows with cells with a borough that is Not assigned, there is <b>NO</b> Not assigned neighborhood. 

In [234]:
missing_data = df.isnull()
missing_data.head(5)
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

PostalCode
False    103
Name: PostalCode, dtype: int64

Borough
False    103
Name: Borough, dtype: int64

Neighbourhood
False    103
Name: Neighbourhood, dtype: int64



In [235]:
df.shape

(103, 3)

The zip code is no longer listed multiple times on the Wiki website. The multiple neighborhoods have already been combined, with a "/". This data cleaning step could be skipped. 

### 4. Load Coordinates

In [236]:
!pip install geocoder
import geocoder #import geocoder

In [237]:
#Load the coordinates from the csv file
df_coordinates = pd.read_csv('http://cocl.us/Geospatial_data/Geospatial_Coordinates.csv')
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [238]:
# rename the column "PostalCode"
df_coordinates.rename(columns={"Postal Code": "PostalCode"}, inplace=True)
df_coordinates.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### 5. Use the csv file to create the new dataframe with coordinates.

In [239]:
df = df.set_index('PostalCode')

In [240]:
df_coordinates = df_coordinates.set_index('PostalCode')
df_coordinates.head()

Latitude  Longitude
PostalCode                      
M1B         43.806686 -79.194353
M1C         43.784535 -79.160497
M1E         43.763573 -79.188711
M1G         43.770992 -79.216917
M1H         43.773136 -79.239476

In [241]:
df_new = pd.concat([df, df_coordinates.reindex(df.index)], axis=1)
df_new.head()

Borough                                 Neighbourhood  \
PostalCode                                                                   
M3A               North York                                     Parkwoods   
M4A               North York                              Victoria Village   
M5A         Downtown Toronto                    Regent Park / Harbourfront   
M6A               North York             Lawrence Manor / Lawrence Heights   
M7A         Downtown Toronto  Queen's Park / Ontario Provincial Government   

             Latitude  Longitude  
PostalCode                        
M3A         43.753259 -79.329656  
M4A         43.725882 -79.315572  
M5A         43.654260 -79.360636  
M6A         43.718518 -79.464763  
M7A         43.662301 -79.389494

In [242]:
df_new = df_new.reset_index()
df_new.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                          Regent Park / Harbourfront  43.654260 -79.360636  
3                   Lawrence Manor / Lawrence Heights  43.718518 -79.464763  
4        Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                     Malvern / Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                    Parkview Hill / Woodbine Gardens  43.706397 -79.309937  
9                           Garden District / Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park / Princess Gardens / Martin Gr...  43.650943 -79.554724

### 6. Explore Toronto

#### Get geocodes of Toronto

In [ ]:
!conda install -c conda-forge geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
!conda install -c conda-forge folium=0.5.0 
import folium # map rendering library

### 7. Explroe neighbourhoods using Foursquare API
Define Foursquare Credentials and Version

In [243]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: WFPGX20PHLXCGUGMROEBZZPVS0XBJQR0N5WS54WKLFRTEE3B
CLIENT_SECRET:P414KZXRB5A2AG0H5RXBXCPG0XVCLPUCLHJAA2COXRFLA44Y


#### Toronto Neighbourhoods
<i> Explore the Neighbourhoods of boroughs that contain the word Toronto</i>

In [244]:
# includes borough that contain the word Toronto
df_new2 = df_new[df_new['Borough'].str.contains('Toronto')].reset_index(drop=True)
df_new2.head()

PostalCode           Borough                                 Neighbourhood  \
0        M5A  Downtown Toronto                    Regent Park / Harbourfront   
1        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District / Ryerson   
3        M5C  Downtown Toronto                                St. James Town   
4        M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

#### Get Toronto geocodes

In [245]:
address = 'Toronto'
geolocator = Nominatim(user_agent='my-application')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [246]:
print('There are {} Postal Codes in the neighbourhoods of the Boroughs named Toronto.'.format(df_new2.shape[0] ))

There are 39 Postal Codes in the neighbourhoods of the Boroughs named Toronto.


#### Explore all neighborhoods in Toronto. 
Extracting data

In [247]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [248]:
toronto_venues = getNearbyVenues(names=df_new2['Neighbourhood'],
                                   latitudes=df_new2['Latitude'],
                                   longitudes=df_new2['Longitude']
                                  )

Regent Park / Harbourfront
Queen's Park / Ontario Provincial Government
Garden District / Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
The Danforth West / Riverdale
Toronto Dominion Centre / Design Exchange
Brockton / Parkdale Village / Exhibition Place
India Bazaar / The Beaches West
Commerce Court / Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park / The Junction South
North Toronto West
The Annex / North Midtown / Yorkville
Parkdale / Roncesvalles
Davisville
University of Toronto / Harbord
Runnymede / Swansea
Moore Park / Summerhill East
Kensington Market / Chinatown / Grange Park
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport
Rosed

#### Size of the dataframe

In [250]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park / Harbourfront               43.65426              -79.360636   
1  Regent Park / Harbourfront               43.65426              -79.360636   
2  Regent Park / Harbourfront               43.65426              -79.360636   
3  Regent Park / Harbourfront               43.65426              -79.360636   
4  Regent Park / Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

#### Number of unique venue categories and number of venues in the neighbourhoods

In [251]:
toronto_venues.groupby(['Neighborhood']).count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton / Parkdale Village / Exhibition Place                         23   
Business reply mail Processing Centre                                  18   
CN Tower / King and Spadina / Railway Lands / H...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             36   
Davisville North                                                        8   
Dufferin / Dovercourt Village                                          19   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                5   
Garden District / Ryerson                                             100   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         23   
India Bazaar / The Beaches West                                        20   
Kensington Market / Chinatown / Grange Park                            55   
Lawrence Park                                                           4   
Little Portugal / Trinity                                              41   
Moore Park / Summerhill East                                            1   
North Toronto West                                                     20   
Parkdale / Roncesvalles                                                13   
Queen's Park / Ontario Provincial Government                           38   
Regent Park / Harbourfront                                             48   
Richmond / Adelaide / King                                             94   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede / Swansea                                                    39   
St. James Town                                                         77   
St. James Town / Cabbagetown                                           46   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West / Rathnelly / South Hill / Fore...                     17   
The Annex / North Midtown / Yorkville                                  22   
The Beaches                                                             4   
The Danforth West / Riverdale                                          43   
Toronto Dominion Centre / Design Exchange                             100   
University of Toronto / Harbord                                        36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton / Parkdale Village / Exhibition Place                          23   
Business reply mail Processing Centre                                   18   
CN Tower / King and Spadina / Railway Lands / H...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              36   
Davisville North                    

In [252]:
print('There are {} uniques venue categories,'.format(len(toronto_venues['Venue Category'].unique())))
print('and {} venues.'.format(len(toronto_venues['Venue'].unique())))

There are 231 uniques venue categories,
and 1054 venues.


### 8. Analyze Each Neighborhood

In [253]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first two column
fixed_columns = [toronto_onehot.columns[-2]] + list(toronto_onehot.columns[:-2])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Women's Store  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0              0        0                   0             0               0   
1              0        0                   0             0               0   
2              0        0                   0             0               0   
3              0        0                   0             0               0   
4              0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               1   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0        0     0                          0            0   
1        0     0                          0            0   
2        0     0                          0            0   
3        0     0                          0            0   
4        0     0                          0    

Size of the new data frame

In [254]:
toronto_onehot.shape

(1623, 230)

#### Group rows by neighborhood and by taking the mean of the frequency of occurence of each category

In [255]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Women's Store  \
0                                         Berczy Park       0.000000   
1      Brockton / Parkdale Village / Exhibition Place       0.000000   
2               Business reply mail Processing Centre       0.000000   
3   CN Tower / King and Spadina / Railway Lands / ...       0.000000   
4                                  Central Bay Street       0.000000   
5                                            Christie       0.000000   
6                                Church and Wellesley       0.000000   
7                     Commerce Court / Victoria Hotel       0.000000   
8                                          Davisville       0.000000   
9                                    Davisville North       0.000000   
10                      Dufferin / Dovercourt Village       0.000000   
11            First Canadian Place / Underground city       0.000000   
12                           Forest Hill North & West       0.000000   
13                          Garden District / Ryerson       0.000000   
14  Harbourfront East / Union Station / Toronto Is...       0.000000   
15                     High Park / The Junction South       0.000000   
16                    India Bazaar / The Beaches West       0.000000   
17        Kensington Market / Chinatown / Grange Park       0.000000   
18                                      Lawrence Park       0.000000   
19                          Little Portugal / Trinity       0.000000   
20                       Moore Park / Summerhill East       0.000000   
21                                 North Toronto West       0.000000   
22                            Parkdale / Roncesvalles       0.000000   
23       Queen's Park / Ontario Provincial Government       0.000000   
24                         Regent Park / Harbourfront       0.000000   
25                         Richmond / Adelaide / King       0.010638   
26                                           Rosedale       0.000000   
27                                           Roselawn       0.000000   
28                                Runnymede / Swansea       0.000000   
29                                     St. James Town       0.000000   
30                       St. James Town / Cabbagetown       0.000000   
31                                     Stn A PO Boxes       0.000000   
32                                    Studio District       0.000000   
33  Summerhill West / Rathnelly / South Hill / For...       0.000000   
34              The Annex / North Midtown / Yorkville       0.000000   
35                                        The Beaches       0.000000   
36                      The Danforth West / Riverdale       0.000000   
37          Toronto Dominion Centre / Design Exchange       0.000000   
38                    University of Toronto / Harbord       0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.058824            0.058824      0.058824        0.117647   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000000            0.000000      0.000000        0.000000   
15  0.000000            0.000000      0.000000        0.000000   
16  0.000000            0.000000      

The grouped by neighborhood new dataframe size

In [256]:
toronto_grouped.shape
print('There are {} neighborhood were returned by Foursquare.'.format(toronto_grouped.shape[0]))
print('There are {} uniques venue categories,'.format(len(toronto_venues['Venue Category'].unique())))
print('and {} venues.'.format(len(toronto_venues['Venue'].unique())))

There are 39 neighborhood were returned by Foursquare.
There are 231 uniques venue categories,
and 1054 venues.


#### Put into a <i>pandas</i> dataframe

Sort the venues in descdnding order.

In [257]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood.

In [258]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing Centre           Pizza Place   
3  CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop              Beer Bar   
1           Coffee Shop             Nightclub        Breakfast Spot   
2            Skate Park      Recording Studio                   Spa   
3        Airport Lounge      Airport Terminal                   Bar   
4    Italian Restaurant                  Café        Sandwich Place   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Seafood Restaurant                Bakery            Restaurant   
1  Performing Arts Venue               Stadium                Bakery   
2     Light Rail Station  Fast Food Restaurant         Auto Workshop   
3               Boutique   Rental Car Location         Boat or Ferry   
4           Burger Joint       Bubble Tea Shop        Ice Cream Shop   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                       Café        Breakfast Spot          Shopping Mall  
1               Intersection    Italian Restaurant           Climbing Gym  
2             Farmers Market            Smoke Shop             Restaurant  
3            Harbor / Marina      Sculpture Garden           Airport Gate  
4  Middle Eastern Restaurant   Fried Chicken Joint        Thai Restaurant

### 9. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [259]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop(['Neighborhood'], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[:] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 4, 0, 2, 0,
       0, 0, 0, 0, 3, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [260]:
neighborhoods_venues_sorted.rename(columns={"Neighborhood": "Neighbourhood"}, inplace=True)

In [261]:
df_new2 = df_new2[['PostalCode', 'Neighbourhood','Borough','Latitude', 'Longitude']]
df_new2.head()

PostalCode                                 Neighbourhood           Borough  \
0        M5A                    Regent Park / Harbourfront  Downtown Toronto   
1        M7A  Queen's Park / Ontario Provincial Government  Downtown Toronto   
2        M5B                     Garden District / Ryerson  Downtown Toronto   
3        M5C                                St. James Town  Downtown Toronto   
4        M4E                                   The Beaches      East Toronto   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [262]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
#toronto_merged["Cluster Labels"] = kmeans.labels_

toronto_merged = df_new2

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

PostalCode                                 Neighbourhood           Borough  \
0        M5A                    Regent Park / Harbourfront  Downtown Toronto   
1        M7A  Queen's Park / Ontario Provincial Government  Downtown Toronto   
2        M5B                     Garden District / Ryerson  Downtown Toronto   
3        M5C                                St. James Town  Downtown Toronto   
4        M4E                                   The Beaches      East Toronto   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0                  Café   
4  43.676357 -79.293031               0                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1      Sushi Restaurant                 Diner    College Auditorium   
2           Coffee Shop                  Café   Japanese Restaurant   
3           Coffee Shop             Gastropub          Cocktail Bar   
4     Health Food Store                   Pub             Wine Shop   

  5th Most Common Venue      6th Most Common Venue  \
0        Breakfast Spot                    Theater   
1         Burrito Place                   Beer Bar   
2            Restaurant  Middle Eastern Restaurant   
3   American Restaurant             Cosmetics Shop   
4          Dance Studio          Electronics Store   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                         Café     Electronics Store           Event Space   
1           Italian Restaurant             Juice Bar        Sandwich Place   
2              Bubble Tea Shop    Italian Restaurant        Cosmetics Shop   
3           Italian Restaurant        Lingerie Store               Theater   
4  Eastern European Restaurant            Donut Shop      Doner Restaurant   

  10th Most Common Venue  
0             Restaurant  
1           Burger Joint  
2         Lingerie Store  
3       Department Store  
4                Dog Run

#### Create Map

In [263]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### 10. Examine Clustsers

#### Cluster 1

In [264]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
0                          Regent Park / Harbourfront               0   
1        Queen's Park / Ontario Provincial Government               0   
2                           Garden District / Ryerson               0   
3                                      St. James Town               0   
4                                         The Beaches               0   
5                                         Berczy Park               0   
6                                  Central Bay Street               0   
7                                            Christie               0   
8                          Richmond / Adelaide / King               0   
9                       Dufferin / Dovercourt Village               0   
10  Harbourfront East / Union Station / Toronto Is...               0   
11                          Little Portugal / Trinity               0   
12                      The Danforth West / Riverdale               0   
13          Toronto Dominion Centre / Design Exchange               0   
14     Brockton / Parkdale Village / Exhibition Place               0   
15                    India Bazaar / The Beaches West               0   
16                    Commerce Court / Victoria Hotel               0   
17                                    Studio District               0   
20                                   Davisville North               0   
22                     High Park / The Junction South               0   
23                                 North Toronto West               0   
24              The Annex / North Midtown / Yorkville               0   
25                            Parkdale / Roncesvalles               0   
26                                         Davisville               0   
27                    University of Toronto / Harbord               0   
28                                Runnymede / Swansea               0   
30        Kensington Market / Chinatown / Grange Park               0   
31  Summerhill West / Rathnelly / South Hill / For...               0   
32  CN Tower / King and Spadina / Railway Lands / ...               0   
34                                     Stn A PO Boxes               0   
35                       St. James Town / Cabbagetown               0   
36            First Canadian Place / Underground city               0   
37                               Church and Wellesley               0   
38              Business reply mail Processing Centre               0   

   1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0            Coffee Shop                  Park                 Bakery   
1            Coffee Shop      Sushi Restaurant                  Diner   
2         Clothing Store           Coffee Shop                   Café   
3                   Café           Coffee Shop              Gastropub   
4                  Trail     Health Food Store                    Pub   
5            Coffee Shop          Cocktail Bar            Cheese Shop   
6            Coffee Shop    Italian Restaurant                   Café   
7          Grocery Store                  Café                   Park   
8            Coffee Shop                  Café             Restaurant   
9               Pharmacy                Bakery              Wine Shop   
10           Coffee Shop              Aquarium                  Hotel   
11                   Bar            Restaurant            Men's Store   
12      Greek Restaurant           Coffee Shop     Italian Restaurant   
13           Coffee Shop                 Hotel                   Café   
14                  Café           Coffee Shop              Nightclub   
15        Sandwich Place  Fast Food Restaurant                   Park   
16           Coffee Shop                  Café             Restaurant   
17                  Café           Coffee Shop              Gastropub   
20                  Park                   Gym         Breakfast Spot   
22    Mexican Restaurant

#### Cluster 2

In [265]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
19      Roselawn               1                Garden             Wine Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19          Dance Studio  Ethiopian Restaurant     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
19               Dog Run    Distribution Center

#### Cluster 3

In [267]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
29  Moore Park / Summerhill East               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29             Wine Shop         Deli / Bodega  Ethiopian Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
29     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center

#### Cluster 4

In [268]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
21  Forest Hill North & West               3      Sushi Restaurant   
33                  Rosedale               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                  Park         Jewelry Store                 Trail   
33            Playground                 Trail          Cupcake Shop   

   5th Most Common Venue        6th Most Common Venue  \
21              Bus Line                    Wine Shop   
33     Electronics Store  Eastern European Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
21  Eastern European Restaurant            Donut Shop      Doner Restaurant   
33                   Donut Shop      Doner Restaurant               Dog Run   

   10th Most Common Venue  
21                Dog Run  
33    Distribution Center

#### Cluster 5

In [269]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, 
                   toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
18  Lawrence Park               4                  Park           Swim School   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18  Construction & Landscaping              Bus Line     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
18  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
18               Dog Run    Distribution Center

## Observation
- In this analysis, it includes 1054 venues of 231 uniques venue categories, in the 39 Boroughs with "Toronto" in the name. 
- The 39 Neighborhoods are clustsered into 5 groups, based on similarity/dissimilarity of the common venues categories. 
- The top 10 most common venue categories in Cluster 1 is a good indicator that these are communities of foodie and coffee lovers. 
- Clustsers 2-5 are rather small in size. It is not as easy to infer dissmiliarities between clusters. Futher evaluation of the model should be performed. 